beacon_node/beacon_chain/src/sync_committee_verification.rs

```diff
@@ -0,0 +1,713 @@
+    /// Returns `Ok(Self)` if the `signed_aggregate` is valid to be (re)published on the gossip
+    /// network.
+    pub fn verify(
+        signed_aggregate: SignedContributionAndProof<T::EthSpec>,
+        chain: &BeaconChain<T>,
+    ) -> Result<Self, Error> {
+        let aggregator_index = signed_aggregate.message.aggregator_index;
+        let contribution = &signed_aggregate.message.contribution;
+        let subcommittee_index = contribution.subcommittee_index as usize;
+
+        // Ensure sync committee contribution is within the MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance.
+        verify_propagation_slot_range(chain, contribution)?;
+
+        // Validate subcommittee index.
+        if contribution.subcommittee_index >= SYNC_COMMITTEE_SUBNET_COUNT {
+            return Err(Error::InvalidSubcommittee {
+                subcommittee_index: contribution.subcommittee_index,
+                subcommittee_size: SYNC_COMMITTEE_SUBNET_COUNT,
+            });
+        }
+
+        // Ensure the valid sync contribution has not already been seen locally.
+        let contribution_root = contribution.tree_hash_root();
+        if chain
+            .observed_sync_contributions
+            .write()
+            .is_known(contribution, contribution_root)
+            .map_err(|e| Error::BeaconChainError(e.into()))?
+        {
+            return Err(Error::SyncContributionAlreadyKnown(contribution_root));
+        }
+
+        // Ensure there has been no other observed aggregate for the given `aggregator_index`.
+        //
+        // Note: do not observe yet, only observe once the sync contribution has been verified.
+        let observed_key =
+            SlotSubcommitteeIndex::new(contribution.slot, contribution.subcommittee_index);
+        match chain
+            .observed_sync_aggregators
+            .read()
+            .validator_has_been_observed(observed_key, aggregator_index as usize)
+        {
+            Ok(true) => Err(Error::AggregatorAlreadyKnown(aggregator_index)),
+            Ok(false) => Ok(()),
+            Err(ObservedAttestersError::ValidatorIndexTooHigh(i)) => {
+                Err(Error::ValidatorIndexTooHigh(i))
+            }
+            Err(e) => Err(BeaconChainError::from(e).into()),
+        }?;
+
+        // Ensure the block being voted for (contribution.beacon_block_root) passes validation.
+        //
+        // This indirectly checks to see if the `contribution.beacon_block_root` is in our fork
+        // choice. Any known, non-finalized, processed block should be in fork choice, so this
+        // check immediately filters out sync committee contributions that attest to a block that
+        // has not been processed.
+        //
+        // Sync committee contributions must be for a known block. If the block is unknown, we
+        // simply drop the sync committee contribution and do not delay consideration for later.
+        verify_head_block_is_known(chain, contribution.beacon_block_root)?;
```

Comment:
We need to implement the changes from [alpha.8](https://github.com/ethereum/eth2.0-specs/releases/tag/v1.1.0-alpha.8), one of which removes this block validity check. The main changes are in this PR https://github.com/ethereum/eth2.0-specs/pull/2491.

I was initially concerned that we might need to make changes to `observed_sync_aggregators` to cache the contributions themselves, however I think this isn't necessary. We can continue using `observed_sync_aggregators` to do the duplicate checking while forwarding the valid contributions immediately to the op pool, which doesn't care whether the block root has been seen previously.

In terms of bounding the size of `observed_sync_aggregators`, we don't explicitly enforce a max size, but it will happen naturally due to the re-ordering of the checks, i.e. validators are only inserted once we've checked that they're part of the relevant sync committee.

Reply:
Made these updates here: https://github.com/sigp/lighthouse/pull/2321/commits/ad32be206da64618961476a1d118afe02ab52cff